### DISCUSSION TITLE
4th gen and digital i/o

### DISCUSSION THREAD

**Discussion by wanthalf** [geoffreybennett/alsa-scarlett-gui#150] [General]
Hi,

First of all, thank you very much for the excellent work!  I don't want to ask The-Infamous-Question (i.e. "when"), I just hope the support for large 4th gen Scarlett is still promising and going on well. The devices are just appearing in local shops and I am looking forward to get one, but I want to be sure they will really be fully supported in Linux before I invest my money.

I have asked Focusrite about their incomplete docs for the 4th gen (and even found contradictory information in the 3rd gen guides) concerning the digital I/O on 18i20 using sample rate of 192kHz. Their final answers indicate that the coax S/PDIF _can_ be used for output (but not for input) at 192kHz though, but (!) the number of outputs is limited to 10 only, so that the S/PDIF and headphone outputs cannot be used independently of the 10 analogue outputs. Can you confirm this?

I suppose that they can be used (all?) as independent outputs at least at 96kHz - possibly with the very same limitations as on the 3rd gen? (i.e. coax and optical S/PDIFs are mutually exclusive options and even the 2 last ADAT channels are excluded by using the coax S/PDIF output)

Or is there any difference in the number of inputs, outputs or mixer in/outs between the 3rd and 4th gen at some of the sampling rates?

Many thanks,
Pavel

**Comment by geoffreybennett** [comment]
Hi @wanthalf,

It *will* be supported. I got the driver most of the way there, sufficient to use, but a bit ugly and not production-ready. For various reasons, I've decided to completely rewrite the driver(*). If you want to be the first 4th Gen 18i20 Linux end-user and don't mind being on the bleeding edge then I'll provide you all the assistance you need to get it going in Linux. Contact me by email and we can discuss.

Regarding digital I/O, for myself I stick to 48kHz because humans can't hear above 20kHz and you just lose features at the higher sample rates. Sorry I can't answer for certain about S/PDIF and ADAT at higher sample rates right now, but I can give you some contradictory information:
- https://userguides.focusrite.com/hc/en-gb/articles/21616107507218-Scarlett-18i20-Specifications (look under "Other I/O" and "input channel order")
- S/PDIF is only up to 96kHz according to the user manual
- S/PDIF is available even at 192kHz according to Focusrite Control 2 ("Mix Sources and ADAT channels are unavailable at sample rates above 96 kHz", and both S/PDIF inputs and outputs can be selected). Although it lets you select playback channels over 10 so I'm not sure that's right.

The limit of 10 is based on the number of USB channels:
```
  Sample Rate     Playback    Capture
 44100,  48000       24          20
 88200,  96000       18          18
176400, 192000       10          10
```
You should be able to route the S/PDIF and headphone outputs independently of the 10 analogue outputs, but you've only got 10 independent USB playback channels.

Also, optical S/PDIF is not currently available. I can find no option for it in Focusrite Control, the v1.1 PDF user guide doesn't say how to set it, and the online user guide https://userguides.focusrite.com/hc/en-gb/articles/21616151554578-Using-Focusrite-Control-2-with-your-Scarlett-18i20#UUID-0a259d33-7381-df5b-a6d5-c82114075531_UUID-c5c2685d-44c5-d6da-c18e-729a8dd02732 says that the feature is "coming soon".

Regards,
Geoffrey.

(*) Reasons: The 4th Gen firmware kept moving the parameters around in the configuration space. The locations are documented internally, but in JSON which I can't parse from the kernel driver. So either the driver needs to be updated for each firmware revision, or I need a userspace helper. The kernel driver is getting rather big/unwieldy, and it still doesn't support _every_ feature. It's also still lacking native stereo/mute/solo support due to no place to store the settings, but I recently discovered where I can put those. The new kernel driver is tiny in comparison, and I'm reimplementing everything in userspace so I can add more features more easily.

**Comment by wanthalf** [comment]
Thanks a lot for confirming many of my suspicions:

Yes, those examples are are some of the contradictory information I came across... :)

The limit of 10/10 USB channels at 192kHz is a limit of the USB UAC protocol, or the USB2 bandwidth or a limit of Focusrite specifically? According to my (somewhat unclear) findings, it must be the last case, since I have read somewhere (about the UAC 2.0) that "As the data rate of High Speed is 40 X Full speed, recording a 60 channel using 24 bits at 96 kHz  (132 Mbit/s) is not a problem." So, that would be 30 channels at 192kHz at least (USB2 is actually 480Mbit, not "just" 132). But if I have "only" 10 independent USB channels available at 192kHz, I just cannot output more independent channels, whatever I do...

Yes, people from Focusrite also made me aware of the "coming soon" state of affairs...

Since you have been working on adding the 4th gen for more than a month(?) and the hardware should't really be that different from the 3rd gen, I expected that you either came across some serious problem with the 4th gen, or that you are rewriting the whole thing for some reason. I am glad to hear it is (just) the latter case! I would be happy to offer you some kind of help, either with implementation or testing, but I am afraid I wouldn't really be able to do anything that you probably couldn't do yourself better and faster. While I have some basic command of C(++), I am not really experienced in low-level programming and especially not "non-ugly and production-ready". But I definitely don't mind experimenting, testing and being on the bleeding edge. That's actually the main reason why I am interested in the audio interface! :)

Best regards,
Pavel

Just FIY: My main interest in multichannel output is actually experimenting with surround sound ("immersive", as they call it nowadays) and possibly also playback of multichannel hi-res audio. But don't take me wrong: I am not the kind of "audiophile" who pretends "hearing" a difference between two power cables! I am aware that I probably wouldn't be able to tell a difference between a bit-perfect FLAC playback and the same recording converted to good quality MP3. And I know (I did test) that my hearing stops somewhere at 15kHz at this age (used to be 21kHz when I was a child) and it will get just worse and worse with the age :( So, I suppose the whole thing is just about a "good feeling" rather than any real effect.

Nevertheless, what I really *do* enjoy is the spacial dimension of sound, where my hearing is (hopefully) not as deteriorated (yet?). But I have to admit that: my equipment does not have more than 10 speakers anyway (and I am not sure I will ever be able to add any more); most of my "immersive" recordings do not have more than 10 channels (at 96kHz) either and are encoded in Auro3D or Dolby Atmos (that I will probably never be able to decode by open source software). But they are slowly being available also in discrete 7.1.4 channel formats (and they are not that much overpriced either, at least not at 88/96kHz). I am also aware that with these audio interfaces I won't be able to play DSD recordings (without conversion to PCM). So, the whole point is rather experimenting, and so I am interested in the theoretical prospects and limits - often just for some hypothetical future applications than actual real need at the very moment.

**Comment by geoffreybennett** [comment]
Hi @wanthalf,

> The limit of 10/10 USB channels at 192kHz is a limit of the USB UAC protocol, or the USB2 bandwidth or a limit of Focusrite specifically?

No idea on that sorry. I just try and implement what the interface makes available :). There could be internal bandwidth limitations in the chipsets that Focusrite uses.

> But if I have "only" 10 independent USB channels available at 192kHz, I just cannot output more independent channels, whatever I do...

Correct.

> ...the hardware should't really be that different from the 3rd gen...

"Shouldn't" :). The underlying protocol is the same, but every different model seems to have something unique about it. The "big" 4th Gen devices have a few of those things. Like an extra onboard processor that has a completely different firmware upgrade procedure.

> ...surround sound...

Oh, that's cool. Yeah, this interface will be fine for 10 channels at 192kHz or 18 channels at 96kHz, presuming the ADAT mode defaults to dual.

Regards,
Geoffrey.

**Comment by wanthalf** [comment]
> "Shouldn't" :)

I know. Some manufacturers even change crucial components in various "revisions" of the very same product, suddenly rendering it useless for any OSS users. The more I have been nervous why is it taking so much time to add support for a few updated models.

> Like an extra onboard processor

Do you refer to the DSP here?

By the way: the DSP is two-channel only and limited to max. 96kHz, if I remember correctly? I suppose there are (currently) no other uses for it other than the two Air functions, right?

So, I will contact you when I get the device. (It will take a week or two at least, I suppose.)
Thank you very much.

**Comment by geoffreybennett** [comment]
> I know. Some manufacturers even change crucial components in various "revisions" of the very same product, suddenly rendering it useless for any OSS users. The more I have been nervous why is it taking so much time to add support for a few updated models.

Focusrite don't seem to have done that so far, as far as I know. At least when the AKM factory burned down and they had to update the Clarett, they created a new product series, even though the products are very similar. That might even be a marketing thing :).

> Do you refer to the DSP here?

No. I'm not very familiar with the architecture, but AFAIK the DSP is XMOS which they been using for a long time. The extra processor in 4th Gen 16i16+ is an ESP. According to https://support.focusrite.com/hc/en-gb/articles/10742398294930-What-is-Air-on-my-interface the Air functionality "is analogue on all our interfaces; there's no DSP taking place". I don't know of anything particularly interesting that you can do with the DSP except in the Vocaster series, which supports a compressor + 5× biquad filters.

> So, I will contact you when I get the device. (It will take a week or two at least, I suppose.)
> Thank you very much.

:+1: You're welcome.

**Comment by wanthalf** [comment]
It seems they only received a few pieces of 18i20 and that's all. Next deliveries promised any time from next week to Christmas. My distributor didn't even give a date yet...

Right, DSP was for "Clip Safe". But apparently also for the new "Air Drive" (while "Air Presence" is analogue, indeed).

What does "ESP" mean? I don't suppose they put in an ESP32 SoC,  however amazing that would (potentially) be.

Meanwhile, I also realised my plans have a serious problem: my AVR only has analog input for 7.1 surround, no more channels. And it seems that more recent devices often don't even have that - whatever price level I look at. Everything done via HDMI inputs. I wouldn't mind getting a better amplifier, but then I would have to swap 10 cables manually (either the speakers or input cables) all the time. Impossible. In addition, you mention that the 4th Gen interfaces cannot adjust their output volume? That would mean I couldn't regulate volume at all if I connected the interface directly to a power amplifier (without its own volume control)!?

**Comment by geoffreybennett** [comment]
Yes, I think it's an ESP32.

"the 4th Gen interfaces cannot adjust their output volume" — that's not correct; they all have an "Output" knob (previous gens labelled it "Monitor") which controls the volume of certain outputs. For the 4th Gen 18i20 that's the Monitor 1 L/R outputs (or Monitor 2 L/R if Alt is elected).

Larger (>4i4) models from previous generations let you select which outputs the Monitor knob controlled (e.g. in your case you'd want to have it control all outputs together), but I don't see how to do that in Focusrite Control 2 yet.

I see at https://userguides.focusrite.com/hc/en-gb/articles/21616151329810-Hardware-Features it describes the knob as: "Main Speaker Output Control ... By default, control the level going to Outputs 1 and 2. ... You can configure Output to control more outputs." but I don't see how to actually do that. I would guess that it's a "coming soon" feature.

Even if you can't configure the volume knob yet, you can control the volume from Linux.

**Comment by wanthalf** [comment]
Software volume as the last resort? That sounds like a terrible waste of dynamics for my perfectionism longing after the "bit perfect playback" :) But if I count correctly, I would still have more than 17 bits left out of the 24 even if I used just 1% of the whole range. That's surely more than my ears (or any reasonably priced amplifier) could deal with.

Shouldn't I also be able to use the mixer for this purpose? It has a limited number of channels - is it still 25in/12out on the 18i20? Or does it also depend on the sampling rate?

So, you say there is some output volume control, but currently only applicable to the first two output channels? And that the confusing documentation seems to promise some unspecified possibility to apply it to "more"(?) channels?

I was hoping you had got better insight into the technical details, since FR promised you some support. So, do you still have to reverse engineer the settings protocol from their Focusrite Control 2 software? (I understand that we will always be limited by the options offered by their firmware's API, of course.)

An ESP32, really? Then I could finally implement the remote OSC control I desire all the time and make it the perfect universal sound device! :D
OK, probably not. They will hardly ever publish the control protocol and surely not the code of their own firmware either.  But it looks like they might be considering implementation of their own remote control to make the interface work as a standalone mixer like the larger MOTU interfaces do? That would be great. Actually, that was at first also my greatest motivation to buy a MOTU - until I heard about the issues people have with them, especially in Linux.

**Comment by geoffreybennett** [comment]
> Shouldn't I also be able to use the mixer for this purpose? It has a limited number of channels - is it still 25in/12out on the 18i20? Or does it also depend on the sampling rate?

The mixer is disabled at 192kHz, but otherwise, it has 12 outputs so that should work. The mixer inputs are not assignable anymore, but there's 43 of them now; one for (almost) every possibility: 24 PCM playback channels, 9 analogue inputs, 2 S/PDIF, 8 ADAT (just no sending mixer outputs back to mixer inputs anymore).

> So, you say there is some output volume control, but currently only applicable to the first two output channels?

That's right. Assigning the output volume control to more channels is "coming soon". The code wasn't ready in time for release day.

> I was hoping you had got better insight into the technical details, since FR promised you some support. So, do you still have to reverse engineer the settings protocol from their Focusrite Control 2 software? (I understand that we will always be limited by the options offered by their firmware's API, of course.)

Focusrite have given me all the support that I've needed so far, which hasn't needed to be very much. I found nearly everything myself after they sent me the devices, just needed a few hints. E.g. when I first looked at them I noticed the new firmware update process, so I put that to the side hoping they could save me some time and just explain it to me. I scheduled a call with one of their developers and while putting together a list of questions for him I ended up figuring it all out in an evening anyway. So I was able to describe the process to him, get confirmation that my guesses were correct, and confirm that I wasn't likely to brick my units with what I was planning to do.

The hard bit with these units wasn't the reverse engineering, but redesigning the driver. Now it's just a simple matter of a little more programming. :crossed_fingers:

> An ESP32, really?

Most of the mentions in the device map are just "ESP", like "ESPBootMode", but there's also "meterDataToESP32", so I think it's at least some sort of ESP32.

> Then I could finally implement the remote OSC control I desire all the time

The XMOS and ESP firmware is encrypted, so that would be "tricky" (to say the least).

>  But it looks like they might be considering implementation of their own remote control to make the interface work as a standalone mixer [...]

I guess that's a possibility. At the moment their remote control android/ios apps talk to Focusrite Control 2, not directly to the device.

Audio interfaces need cables anyway, so I don't think it'd be too bad to bolt a tiny Linux box to the Scarlett, run an ALSA-OSC bridge on it and have an almost-perfect standalone mixer/universal sound device. You could even support Ethernet connectivity then :).

**Comment by geoffreybennett** [comment]
Hi @wanthalf,

An update for you!

My FCP driver has been accepted upstream and almost all the pieces are ready for public release:
- https://github.com/geoffreybennett/linux-fcp
- https://github.com/geoffreybennett/fcp-support
- https://github.com/geoffreybennett/alsa-scarlett-gui g4-big branch

I found that the limit of 10 channels at 192kHz is a USB limit:
https://thewelltemperedcomputer.com/Lib/Fundamentals-of-USB-Audio_1.0.pdf

And switching the digital I/O mode between S/PDIF RCA, S/PDIF Optical, and Dual ADAT is now possible from Linux.

Regards,
Geoffrey.

**Reply by wanthalf** [reply]
Hi Geoffrey,

congratulations! I am watching and I am happy for that, despite the fact that my own situation did a turn around and I ... well, became a "traitor". After two weeks my order was canceled by the seller as "undeliverable" - while ironically, the 18i20 suddenly appeared "in stock" just some three days later in their e-shop! However, in between... independent [measurements](https://www.audiosciencereview.com/forum/index.php?threads/focusrite-scarlett-2i2-4th-gen-interface-review.58790/) of the new Scarlett 2i2 were published on ASR, which put me in doubt due to some strange (possibly insignificant, but still...) quirks of the new Scarletts. So, I started to reconsider getting a MOTU UL mk5 again and just at the moment discovered that some young musician was selling one second-hand for a good price... almost in my own street, by chance. So I was able to test it myself and while it doesn't offer the luxury of mixer support in ALSA, their mixer control application (unofficially) actually runs even under Linux or Android and... actually already supports all the "coming soon"-features of the Focusrite Gen4, out of the box.

What surprised me more is the lack of support for multichannel (>8) playback in any reasonable Linux player or library - beyond the DAWs at the top level end and the ALSA play utility at the lowest level end... but that's another story. 

Anyway, I will continue watching the Focusrite products and your software and envy the level of support they get in Linux - thanks to you! 

Congratulations and many thanks for your great effort - in the name of the whole Linux musician and audio community!
Pavel

**Comment by geoffreybennett** [comment]
Thanks for the update and your kind words! Your decision makes sense — I had a look at the manual for that thing and it's got some features that aren't on any Scarlett or Clarett (EQ, gate, compressor, reverb). Not as many mic inputs as the 18i20 (it's more like the 16i16 in that regard), and only one headphone output, but as much total I/O as the 18i20.

I'm glad you found a solution that works for you. I'll close this discussion now, but feel free to open another if you have any questions in the future.

**Reply by wanthalf** [reply]
Right, if I needed 8 mic preamps,  I would need to get something else. But the number of outputs is more important now.

By the way, the device's mixer is controlled via a websocket port on a virtual USB NIC created by the card (and then it can also work  from any remote network device via a proxy on the connected computer). In addition, the full javascript sources of MOTU's control application are available, so that one should be able to document the binary protocol and control it by any custom means just by studying the code. Somehow, no one has yet found the desire or time to do that concisely - except for a few single codes extracted by some user for simple one-shot tasks like (un)muting the output from a python script. I am afraid I won't find the time (or real need) to do the job either.  But  in principle, it should be well doable and feels tempting. I wonder if such kind of control over the network stack could also be translated into an ALSA driver.

**Comment by geoffreybennett** [comment]
Definitely possible.  I see https://github.com/alexanderson1993/Motu-Control which says it's for AVB devices, but presumably your device is very similar.

It would just be a simple matter of programming to build a bridge between the MOTU interface and ALSA controls. A very similar architecture to what I'm doing with [`fcp-server`](https://github.com/geoffreybennett/fcp-support).

**Comment by wanthalf** [comment]
I think the AVB supported network control using OSC commands directly - and that seems to be implemented here. It actually had a physical ethernet port and a full-fledged controller (well, it was "AVB"). That was my dream until I noticed people reporting serious audio issues with Linux after some firmware update. If I got it correctly, MOTU claimed it was the Linux implementation of UAC2 which was broken. Anyway, later it went out of production completely.

The "mk5"  and its bigger sibling 828 do not support AVB nor OSC, but they can be controlled using their own binary protocol via the emulated USB NIC. I wish they would publish at least a translator from OSC to the binary code, if they can't/won't implement it on the board anymore. But we can be happy one MOTU developer unofficially published the Electron-based control application [adapted to Linux](https://linuxmusicians.com/viewtopic.php?p=164679#p164679) environment on his own, and the JS/HTML source is actually freely available. Better than pure reverse engineering and definitely better than nothing.
